src/Reflex/Backend/Socket.hs

```diff
@@ -1,192 +1,190 @@
+
+-- | Holds the socket to wire into the FRP network, and events that
+-- drive it.
+data SocketConfig t = SocketConfig
+  { _scInitSocket :: NS.Socket
+    -- ^ Socket to wrap.
+  , _scMaxRx :: Int
+    -- ^ Maximum number of bytes to read at a time.
+  , _scSend :: Event t ByteString
+    -- ^ Data to send out on this socket.
+  , _scClose :: Event t ()
+    -- ^ Ask to close the socket. The socket will stop trying to
+    -- receive data (and the '_sReceive' event will stop firing), and
+    -- the socket will be "drained": future events on '_scSend' will
+    -- be ignored, and it will close after writing all pending data.
+    -- If '_scSend' and '_scClose' fire in the same frame, the data
+    -- will nevertheless be queued for sending.
   }
 
-makeLenses ''SocketConfig
-
-data SocketError
-  = SocketIOException IOException
-  | SocketDecodeError String
-  deriving (Eq, Show)
-
-data Socket t b =
-  Socket {
-    _sReceive :: Event t b
-  , _sOpen    :: Event t ()
-  , _sError   :: Event t SocketError
-  , _sClosed  :: Event t ()
+$(makeLenses ''SocketConfig)
+
+-- | Events produced by an active socket.
+data Socket t = Socket
+  { _sReceive :: Event t ByteString
+    -- ^ Data has arrived.
+  , _sOpen :: Event t ()
+    -- ^ The socket has opened, and its receive/send loops are running.
+  , _sClose :: Event t ()
+    -- ^ The socket has closed. This will fire exactly once when the
+    -- socket closes for any reason, including if your '_scClose'
+    -- event fires, the other end disconnects, or if the socket closes
+    -- in response to a caught exception.
+  , _sError :: Event t IOException
+    -- ^ An exception occurred. Treat the socket as closed after you
+    -- see this. If the socket was open, you will see the '_sClosed'
+    -- event fire as well, but not necessarily in the same frame.
   }
 
-makeLenses ''Socket
-
-socket ::
-  forall t m a b.
-  ( Reflex t
-  , PerformEvent t m
-  , PostBuild t m
-  , TriggerEvent t m
-  , MonadIO (Performable m)
-  , MonadIO m
-  , CanEncode a
-  , CanDecode b
-  ) =>
-  SocketConfig t a ->
-  m (Socket t b)
-socket (SocketConfig initSock mxRx eTx eClose) = mdo
+$(makeLenses ''Socket)
+
+data SocketState
+  = Open
+    -- ^ Data flows in both directions
+  | Draining
+    -- ^ We've been asked to close, but will transmit all pending data
+    -- first (and not accept any more)
+  | Closed
+    -- ^ Hard close. Don't transmit pending data.
+
+-- | Wire a socket into the FRP network. You will likely use this to
+-- attach events to a socket that you just connected (from 'connect'),
+-- or a socket that you just accepted (from the '_aAcceptSocket' event
+-- you got when you called 'accept').
+socket
+  :: forall t m.
+     ( Reflex t
+     , PerformEvent t m
+     , PostBuild t m
+     , TriggerEvent t m
+     , MonadIO (Performable m)
+     , MonadIO m
+     )
+  => SocketConfig t
+  -> m (Socket t)
+socket (SocketConfig sock maxRx eTx eClose) = do
```

Comment:
I am a lot more confident in this now, but I think it needs someone else to think _hard_ about it as well.